torch/nativert/graph/GraphSignature.cpp

```diff
@@ -0,0 +1,509 @@
+  // Retrieve the argument name and spec tag name
+  std::string argName;
+  std::string tagName;
+  switch (inputSpec.tag()) {
+    case torch::_export::InputSpec::Tag::PARAMETER:
+      argName = inputSpec.get_parameter().get_arg().get_name();
+      tagName = "PARAMETER";
+      break;
+    case torch::_export::InputSpec::Tag::BUFFER:
+      argName = inputSpec.get_buffer().get_arg().get_name();
+      tagName = "BUFFER";
+      break;
+    case torch::_export::InputSpec::Tag::TENSOR_CONSTANT:
+      argName = inputSpec.get_tensor_constant().get_arg().get_name();
+      tagName = "TENSOR_CONSTANT";
+      break;
+    case torch::_export::InputSpec::Tag::CUSTOM_OBJ:
+      tagName = "CUSTOM_OBJ";
+      argName = inputSpec.get_custom_obj().get_arg().get_name();
+      break;
+    case torch::_export::InputSpec::Tag::USER_INPUT:
+      tagName = "USER_INPUT";
+      if (inputSpec.get_user_input().get_arg().tag() ==
+          torch::_export::Argument::Tag::AS_TENSOR) {
+        argName =
+            inputSpec.get_user_input().get_arg().get_as_tensor().get_name();
+      } else if (
+          inputSpec.get_user_input().get_arg().tag() ==
+          torch::_export::Argument::Tag::AS_CUSTOM_OBJ) {
+        argName =
+            inputSpec.get_user_input().get_arg().get_as_custom_obj().get_name();
+      } else {
+        TORCH_CHECK(false, "Unsupported USER_INPUT argument type.");
+      }
+      break;
+    case torch::_export::InputSpec::Tag::CONSTANT_INPUT:
+      argName = inputSpec.get_constant_input().get_name();
+      tagName = "CONSTANT_INPUT";
+      break;
+    case torch::_export::InputSpec::Tag::TOKEN:
+      TORCH_CHECK(false, "Token inputs not implemented yet.");
+    default:
+      TORCH_CHECK(false, "Unknown InputSpec tag encountered.");
+  }
+  return std::make_pair(argName, tagName);
+}
+
+void checkInputOrders(
+    const std::vector<torch::_export::InputSpec>& inputSpecs) {
+  // Map each tag to its index in the expected order
+  static constexpr std::pair<torch::_export::InputSpec::Tag, uint32_t>
+      tagOrderArray[] = {
+          {torch::_export::InputSpec::Tag::TOKEN, 0},
+          {torch::_export::InputSpec::Tag::PARAMETER, 1},
+          {torch::_export::InputSpec::Tag::BUFFER, 2},
+          {torch::_export::InputSpec::Tag::TENSOR_CONSTANT, 3},
+          {torch::_export::InputSpec::Tag::CUSTOM_OBJ, 4}};
+  uint32_t currentOrderIndex = 0;
+  bool seenNonPersistentBuffer = false;
+  for (const auto& inputSpec : inputSpecs) {
+    if (inputSpec.tag() == torch::_export::InputSpec::Tag::USER_INPUT ||
+        inputSpec.tag() == torch::_export::InputSpec::Tag::CONSTANT_INPUT) {
+      continue;
+    }
+    uint32_t tagIndex = static_cast<uint32_t>(-1);
+    for (const auto& [tag, index] : tagOrderArray) {
+      if (tag == inputSpec.tag()) {
+        tagIndex = index;
+        break;
+      }
+    }
+    TORCH_CHECK(
+        tagIndex != static_cast<uint32_t>(-1),
+        "Unknown InputSpec tag encountered.");
```

Comment:
it would probably be simpler to use std::find for this after switching to std::array